File: src/caps.ts
export const CAPTURE_BYTE_LIMIT = 8 * 1024 * 1024;
/**
 * Raw claude session .jsonl is dominated by tool output the built-in extractor
 * discards, so the transcript cap applies to the extracted text. This only
 * bounds how large a raw session file the extractor will load.
 */
export const TRANSCRIPT_FILE_BYTE_LIMIT = 32 * CAPTURE_BYTE_LIMIT;
/**
 * A record is buffered whole before its type is known, so this bounds the memory
 * one JSONL record can hold. One record's extracted text can approach the 8 MiB
 * transcript cap and JSON escaping plus non-text blocks multiply the raw size,
 * so 4x leaves headroom for legitimate records.
 */
export const TRANSCRIPT_RECORD_BYTE_LIMIT = 4 * CAPTURE_BYTE_LIMIT;
export const HWF_ENV_BYTE_LIMIT = 24 * 1024;
/**
 * herdr agent.prompt silently drops ~21KB+ bodies; stay under this with a margin.
 * Oversized prompts are written to a run-owned file and replaced by a short pointer.
 */
export const AGENT_PROMPT_BYTE_LIMIT = 16 * 1024;

export class CaptureLimitError extends Error {
  readonly source: string;
  readonly bytes: number;
  readonly limit: number;

  constructor(source: string, bytes: number, limit: number = CAPTURE_BYTE_LIMIT) {
    super(`${source} exceeded ${limit} byte limit (${bytes} bytes)`);
    this.name = "CaptureLimitError";
    this.source = source;
    this.bytes = bytes;
    this.limit = limit;
  }
}

export function assertUnderCaptureCap(source: string, text: string): void {
  const bytes = Buffer.byteLength(text);
  if (bytes > CAPTURE_BYTE_LIMIT) throw new CaptureLimitError(source, bytes);
}

function assertUnderHwfEnvCap(source: string, text: string): void {
  const bytes = Buffer.byteLength(text);
  if (bytes > HWF_ENV_BYTE_LIMIT) throw new CaptureLimitError(source, bytes, HWF_ENV_BYTE_LIMIT);
}

/**
 * Model the byte size of the generated `HWF_*` environment block for the cap check only.
 * `buildHwfEnv` in `src/engine/command.ts` builds the real environment.
 */
function formatHwfEnvBlock(values: Record<string, string>): string {
  return Object.entries(values)
    .map(([name, value]) => `HWF_${name}=${value}`)
    .join("\n");
}

export function assertHwfEnvValues(source: string, values: Record<string, string>): void {
  assertUnderHwfEnvCap(source, formatHwfEnvBlock(values));
}

File: internal/caps/caps.go
// Package caps holds the byte caps and their guards. If data is more than a cap,
// the check fails with the source and the limit. The output is not truncated.
package caps

import (
	"fmt"
	"maps"
	"slices"
	"strings"
)

const CaptureByteLimit = 8 * 1024 * 1024

// TranscriptFileByteLimit is the maximum size of a raw claude session file that
// the built-in extractor loads. The transcript cap applies to extracted text.
const TranscriptFileByteLimit = 32 * CaptureByteLimit

// TranscriptRecordByteLimit is the maximum memory for one buffered JSONL record.
// Extracted text of one record can be almost the capture cap. JSON escaping and
// non-text blocks multiply the raw size, so 4x supplies remaining capacity.
const TranscriptRecordByteLimit = 4 * CaptureByteLimit

const HwfEnvByteLimit = 24 * 1024

// AgentPromptByteLimit stays less than the ~21KB body size that herdr
// agent.prompt discards with no message. Oversized prompts write to a run-owned file.
const AgentPromptByteLimit = 16 * 1024

// CaptureLimitError identifies a source that is more than its byte limit.
type CaptureLimitError struct {
	Source string
	Bytes  int
	Limit  int
}

func (e *CaptureLimitError) Error() string {
	return fmt.Sprintf("%s exceeded %d byte limit (%d bytes)", e.Source, e.Limit, e.Bytes)
}

// AssertUnderCaptureCap fails if text is more than the shared capture cap.
func AssertUnderCaptureCap(source, text string) error {
	if len(text) > CaptureByteLimit {
		return &CaptureLimitError{Source: source, Bytes: len(text), Limit: CaptureByteLimit}
	}
	return nil
}

// formatHwfEnvBlock calculates the byte size of the generated HWF_* environment
// block. The function is only for the cap check.
func formatHwfEnvBlock(values map[string]string) string {
	var b strings.Builder
	for _, name := range slices.Sorted(maps.Keys(values)) {
		fmt.Fprintf(&b, "HWF_%s=%s\n", name, values[name])
	}
	return strings.TrimSuffix(b.String(), "\n")
}

// AssertHwfEnvValues fails if the generated HWF_* environment block is more
// than its cap.
func AssertHwfEnvValues(source string, values map[string]string) error {
	block := formatHwfEnvBlock(values)
	if len(block) > HwfEnvByteLimit {
		return &CaptureLimitError{Source: source, Bytes: len(block), Limit: HwfEnvByteLimit}
	}
	return nil
}
